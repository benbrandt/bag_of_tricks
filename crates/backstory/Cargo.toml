[package]
name = "backstory"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
