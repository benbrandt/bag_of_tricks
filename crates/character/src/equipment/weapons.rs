use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter};

#[derive(Clone, Debug, Deserialize, Display, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub(crate) enum WeaponCategory {
    Simple,
    Martial,
}

#[derive(Clone, Debug, Deserialize, Display, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub(crate) enum WeaponClassification {
    Melee,
    Ranged,
}

#[allow(dead_code)]
#[derive(
    Clone, Copy, Debug, Deserialize, Display, EnumIter, Eq, Ord, PartialEq, PartialOrd, Serialize,
)]
pub(crate) enum WeaponType {
    Battleaxe,
    Blowgun,
    Club,
    #[strum(serialize = "Hand crossbow")]
    CrossbowHand,
    #[strum(serialize = "Heavy crossbow")]
    CrossbowHeavy,
    #[strum(serialize = "Light crossbow")]
    CrossbowLight,
    Dagger,
    Dart,
    Flail,
    Glaive,
    Greataxe,
    Greatsword,
    Greatclub,
    Halberd,
    Handaxe,
    Javelin,
    Lance,
    Longbow,
    Longsword,
    #[strum(serialize = "Light hammer")]
    LightHammer,
    Mace,
    Maul,
    Morningstar,
    Net,
    Pike,
    Quarterstaff,
    Rapier,
    Scimitar,
    Shortbow,
    Shortsword,
    Sickle,
    Sling,
    Spear,
    Trident,
    #[strum(serialize = "War pick")]
    WarPick,
    Warhammer,
    Whip,
}

impl WeaponType {
    pub(crate) fn category(self) -> WeaponCategory {
        match self {
            Self::Club
            | Self::Dagger
            | Self::Greatclub
            | Self::Handaxe
            | Self::Javelin
            | Self::LightHammer
            | Self::Mace
            | Self::Quarterstaff
            | Self::Sickle
            | Self::Spear
            | Self::CrossbowLight
            | Self::Dart
            | Self::Shortbow
            | Self::Sling => WeaponCategory::Simple,
            Self::Battleaxe
            | Self::Flail
            | Self::Glaive
            | Self::Greataxe
            | Self::Greatsword
            | Self::Halberd
            | Self::Lance
            | Self::Longsword
            | Self::Maul
            | Self::Morningstar
            | Self::Pike
            | Self::Rapier
            | Self::Scimitar
            | Self::Shortsword
            | Self::Trident
            | Self::WarPick
            | Self::Warhammer
            | Self::Whip
            | Self::Blowgun
            | Self::CrossbowHand
            | Self::CrossbowHeavy
            | Self::Longbow
            | Self::Net => WeaponCategory::Martial,
        }
    }

    pub(crate) fn classification(self) -> WeaponClassification {
        match self {
            Self::Club
            | Self::Dagger
            | Self::Greatclub
            | Self::Handaxe
            | Self::Javelin
            | Self::LightHammer
            | Self::Mace
            | Self::Quarterstaff
            | Self::Sickle
            | Self::Spear
            | Self::Battleaxe
            | Self::Flail
            | Self::Glaive
            | Self::Greataxe
            | Self::Greatsword
            | Self::Halberd
            | Self::Lance
            | Self::Longsword
            | Self::Maul
            | Self::Morningstar
            | Self::Pike
            | Self::Rapier
            | Self::Scimitar
            | Self::Shortsword
            | Self::Trident
            | Self::WarPick
            | Self::Warhammer
            | Self::Whip => WeaponClassification::Melee,
            Self::CrossbowLight
            | Self::Dart
            | Self::Shortbow
            | Self::Sling
            | Self::Blowgun
            | Self::CrossbowHand
            | Self::CrossbowHeavy
            | Self::Longbow
            | Self::Net => WeaponClassification::Ranged,
        }
    }
}

/// Types of ammunition available
#[derive(Clone, Debug, Deserialize, Display, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub(crate) enum AmmunitionType {
    Arrows,
    #[strum(serialize = "Blowgun needles")]
    BlowgunNeedles,
    #[strum(serialize = "Crossbow bolts")]
    CrossbowBolts,
    #[strum(serialize = "Sling bullets")]
    SlingBullets,
}

// /// Ammunition type and quantity
// struct Ammunition(AmmunitionType, u8);
