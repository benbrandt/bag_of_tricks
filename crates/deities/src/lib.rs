#![deny(clippy::all)]
#![warn(clippy::pedantic)]

mod dragon;
mod dragonlance;
mod dwarven;
mod eberron;
mod elven;
mod forgotten_realms;
mod giant;
mod gnomish;
mod goblinoid;
mod greyhawk;
mod halfling;
mod historical;
mod kobold;
mod lizardfolk;
mod orc;

use std::fmt;

use alignment::{Alignment, Attitude, Morality};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::Display;

use self::{
    dragon::Dragon,
    dragonlance::Dragonlance,
    dwarven::{Duergar, Dwarven},
    eberron::Eberron,
    elven::{Drow, Elven},
    forgotten_realms::ForgottenRealms,
    giant::Giant,
    gnomish::Gnomish,
    goblinoid::{Bugbear, Goblin},
    greyhawk::Greyhawk,
    halfling::Halfling,
    historical::{Celtic, Egyptian, Greek, Norse},
    kobold::Kobold,
    lizardfolk::Lizardfolk,
    orc::Orc,
};

#[derive(Copy, Clone, Deserialize, Display, PartialEq, Serialize)]
pub enum Domain {
    Arcana,
    Death,
    Forge,
    Grave,
    Knowledge,
    Life,
    Light,
    Nature,
    Tempest,
    Trickery,
    War,
}

#[derive(Clone, Deserialize, Serialize)]
pub struct Deity<'a> {
    pub name: &'a str,
    pub titles: Vec<&'a str>,
    pub alignment: Alignment,
    pub domains: Vec<Domain>,
    pub symbols: Vec<&'a str>,
}

impl Deity<'_> {
    fn weight(&self, attitude_influences: &[Attitude], morality_influences: &[Morality]) -> f64 {
        self.alignment
            .weight(attitude_influences, morality_influences)
    }
}

impl<'a> fmt::Display for Deity<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "CHOSEN DEITY: {}", self.name)?;
        writeln!(f, "Titles: {}", self.titles.join("; "))?;
        writeln!(f, "Alignment: {}", self.alignment)?;
        writeln!(
            f,
            "Domains: {}",
            self.domains
                .iter()
                .map(|d| format!("{}", d))
                .collect::<Vec<_>>()
                .join(", ")
        )?;
        writeln!(f, "Symbols: {}", self.symbols.join(", "))
    }
}

pub(crate) trait Deities<'a> {
    fn deities() -> Vec<Deity<'a>>;
}

#[derive(Copy, Clone, Deserialize, Display, PartialEq, Serialize)]
pub enum Pantheon {
    Bugbear,
    Celtic,
    Dragon,
    Dragonlance,
    Drow,
    Duergar,
    Dwarven,
    Eberron,
    Egyptian,
    Elven,
    #[strum(serialize = "Forgotten Realms")]
    ForgottenRealms,
    Giant,
    Gnomish,
    Goblin,
    Greek,
    Greyhawk,
    Halfling,
    Kobold,
    Lizardfolk,
    Norse,
    Orc,
    None,
}

impl Pantheon {
    fn all_deities<'a>(self) -> Vec<Deity<'a>> {
        match self {
            Self::Bugbear => Bugbear::deities(),
            Self::Celtic => Celtic::deities(),
            Self::Dragon => Dragon::deities(),
            Self::Dragonlance => Dragonlance::deities(),
            Self::Drow => Drow::deities(),
            Self::Dwarven => Dwarven::deities(),
            Self::Duergar => Duergar::deities(),
            Self::Eberron => Eberron::deities(),
            Self::Egyptian => Egyptian::deities(),
            Self::Elven => Elven::deities(),
            Self::ForgottenRealms => ForgottenRealms::deities(),
            Self::Giant => Giant::deities(),
            Self::Gnomish => Gnomish::deities(),
            Self::Goblin => Goblin::deities(),
            Self::Greek => Greek::deities(),
            Self::Greyhawk => Greyhawk::deities(),
            Self::Halfling => Halfling::deities(),
            Self::Kobold => Kobold::deities(),
            Self::Lizardfolk => Lizardfolk::deities(),
            Self::Norse => Norse::deities(),
            Self::Orc => Orc::deities(),
            Self::None => vec![],
        }
    }

    #[must_use]
    pub fn deities<'a>(self, domain: Option<Domain>) -> Vec<Deity<'a>> {
        domain.map_or_else(
            || self.all_deities(),
            |d| {
                self.all_deities()
                    .into_iter()
                    .filter(|deity| deity.domains.contains(&d))
                    .collect::<Vec<_>>()
            },
        )
    }

    /// # Panics
    ///
    /// Will panic if no Pantheons available
    pub fn choose(
        rng: &mut impl Rng,
        addl_pantheons: Vec<Self>,
        domain: Option<Domain>,
        required: bool,
    ) -> Self {
        let mut options = vec![
            Self::ForgottenRealms,
            Self::Celtic,
            Self::Dragonlance,
            Self::Eberron,
            Self::Egyptian,
            Self::Greek,
            Self::Greyhawk,
            Self::Norse,
        ];
        options.extend(addl_pantheons);
        if !required {
            options.push(Self::None);
        }
        options.dedup();
        *domain
            .map_or(options.clone(), |_| {
                options
                    .into_iter()
                    .filter(|p| p == &Self::None || !p.deities(domain).is_empty())
                    .collect::<Vec<_>>()
            })
            .choose_weighted(rng, |p| p.weight())
            .unwrap()
    }

    pub fn choose_deity<'a>(
        self,
        rng: &mut impl Rng,
        attitude_influences: &[Attitude],
        morality_influences: &[Morality],
        domain: Option<Domain>,
    ) -> Option<Deity<'a>> {
        self.deities(domain)
            .choose_weighted(rng, |d| d.weight(attitude_influences, morality_influences))
            .ok()
            .cloned()
    }

    #[must_use]
    pub fn weight(self) -> f64 {
        match self {
            Self::Bugbear
            | Self::Dragon
            | Self::Drow
            | Self::Duergar
            | Self::Dwarven
            | Self::Elven
            | Self::ForgottenRealms
            | Self::Giant
            | Self::Gnomish
            | Self::Goblin
            | Self::Halfling
            | Self::Kobold
            | Self::Lizardfolk
            | Self::Orc
            | Self::None => 100.0,
            // Require some sort of plane shift for your character to end up in the forgotten realms
            Self::Celtic
            | Self::Dragonlance
            | Self::Eberron
            | Self::Egyptian
            | Self::Greek
            | Self::Greyhawk
            | Self::Norse => 1.0,
        }
    }
}

pub trait Pantheons {
    fn addl_pantheons(&self) -> Vec<Pantheon> {
        vec![]
    }

    fn deity_required(&self) -> bool {
        false
    }
}
