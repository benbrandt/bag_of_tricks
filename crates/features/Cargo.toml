[package]
name = "features"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2018"
resolver = "2"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
citation = { path = "../citation" }
serde = { version = "1", features = ["derive"] }

[dev-dependencies]
insta = "1"
serde_yaml = "0.8"
