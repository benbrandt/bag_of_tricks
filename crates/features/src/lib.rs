use std::fmt;

use citation::Citation;
use serde::{Deserialize, Serialize};

/// A feature or trait a character has.
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Feature {
    /// Name of the feature or trait.
    pub title: &'static str,
    /// Citation for where more information about this feature is available.
    pub citation: Citation,
}

impl fmt::Display for Feature {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} ({})", self.title, self.citation)
    }
}

/// Trait for objects that provide features to a character.
pub trait Features {
    // Return a list of features this thing provides
    fn features(&self) -> Vec<Feature> {
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use citation::Book;

    #[test]
    fn test_display() {
        let feature = Feature {
            title: "Title",
            citation: Citation(Book::Phb, 1),
        };
        insta::assert_snapshot!(format!("{}", feature));
    }
}
