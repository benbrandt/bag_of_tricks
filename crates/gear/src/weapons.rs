use dice_roller::{Die, RollCmd};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter};

#[derive(
    Clone, Debug, Deserialize, Display, EnumIter, Eq, Ord, PartialEq, PartialOrd, Serialize,
)]
pub enum WeaponCategory {
    Simple,
    Martial,
}

#[derive(Clone, Debug, Deserialize, Display, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub enum WeaponClassification {
    Melee,
    Ranged,
}

#[derive(Clone, Debug, Deserialize, Display, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub enum WeaponProperty {
    Ammunition(Ammunition, u16, u16),
    Finesse,
    Heavy,
    Light,
    Loading,
    Reach,
    Special,
    Thrown(u16, u16),
    #[strum(serialize = "Two-Handed")]
    TwoHanded,
    Versatile(RollCmd),
}

#[allow(dead_code)]
#[derive(
    Clone, Copy, Debug, Deserialize, Display, EnumIter, Eq, Ord, PartialEq, PartialOrd, Serialize,
)]
pub enum Weapon {
    Battleaxe,
    Blowgun,
    Club,
    #[strum(serialize = "Hand crossbow")]
    CrossbowHand,
    #[strum(serialize = "Heavy crossbow")]
    CrossbowHeavy,
    #[strum(serialize = "Light crossbow")]
    CrossbowLight,
    Dagger,
    Dart,
    Flail,
    Glaive,
    Greataxe,
    Greatsword,
    Greatclub,
    Halberd,
    Handaxe,
    Javelin,
    Lance,
    Longbow,
    Longsword,
    #[strum(serialize = "Light hammer")]
    LightHammer,
    Mace,
    Maul,
    Morningstar,
    Net,
    Pike,
    Quarterstaff,
    Rapier,
    Scimitar,
    Shortbow,
    Shortsword,
    Sickle,
    Sling,
    Spear,
    Trident,
    #[strum(serialize = "War pick")]
    WarPick,
    Warhammer,
    Whip,
}

impl Weapon {
    pub fn category(self) -> WeaponCategory {
        match self {
            Self::Club
            | Self::Dagger
            | Self::Greatclub
            | Self::Handaxe
            | Self::Javelin
            | Self::LightHammer
            | Self::Mace
            | Self::Quarterstaff
            | Self::Sickle
            | Self::Spear
            | Self::CrossbowLight
            | Self::Dart
            | Self::Shortbow
            | Self::Sling => WeaponCategory::Simple,
            Self::Battleaxe
            | Self::Flail
            | Self::Glaive
            | Self::Greataxe
            | Self::Greatsword
            | Self::Halberd
            | Self::Lance
            | Self::Longsword
            | Self::Maul
            | Self::Morningstar
            | Self::Pike
            | Self::Rapier
            | Self::Scimitar
            | Self::Shortsword
            | Self::Trident
            | Self::WarPick
            | Self::Warhammer
            | Self::Whip
            | Self::Blowgun
            | Self::CrossbowHand
            | Self::CrossbowHeavy
            | Self::Longbow
            | Self::Net => WeaponCategory::Martial,
        }
    }

    pub fn classification(self) -> WeaponClassification {
        match self {
            Self::Club
            | Self::Dagger
            | Self::Greatclub
            | Self::Handaxe
            | Self::Javelin
            | Self::LightHammer
            | Self::Mace
            | Self::Quarterstaff
            | Self::Sickle
            | Self::Spear
            | Self::Battleaxe
            | Self::Flail
            | Self::Glaive
            | Self::Greataxe
            | Self::Greatsword
            | Self::Halberd
            | Self::Lance
            | Self::Longsword
            | Self::Maul
            | Self::Morningstar
            | Self::Pike
            | Self::Rapier
            | Self::Scimitar
            | Self::Shortsword
            | Self::Trident
            | Self::WarPick
            | Self::Warhammer
            | Self::Whip => WeaponClassification::Melee,
            Self::CrossbowLight
            | Self::Dart
            | Self::Shortbow
            | Self::Sling
            | Self::Blowgun
            | Self::CrossbowHand
            | Self::CrossbowHeavy
            | Self::Longbow
            | Self::Net => WeaponClassification::Ranged,
        }
    }

    pub fn properties(&self) -> Vec<WeaponProperty> {
        match self {
            Self::Battleaxe => vec![WeaponProperty::Versatile(RollCmd(1, Die::D10))],
            Self::Blowgun => vec![
                WeaponProperty::Ammunition(Ammunition::BlowgunNeedles, 25, 100),
                WeaponProperty::Loading,
            ],
            Self::Club => vec![WeaponProperty::Light],
            Self::CrossbowHand => vec![
                WeaponProperty::Ammunition(Ammunition::CrossbowBolts, 30, 120),
                WeaponProperty::Light,
                WeaponProperty::Loading,
            ],
            Self::CrossbowHeavy => vec![
                WeaponProperty::Ammunition(Ammunition::CrossbowBolts, 100, 400),
                WeaponProperty::Heavy,
                WeaponProperty::Loading,
                WeaponProperty::TwoHanded,
            ],
            Self::CrossbowLight => vec![
                WeaponProperty::Ammunition(Ammunition::CrossbowBolts, 80, 320),
                WeaponProperty::Loading,
                WeaponProperty::TwoHanded,
            ],
            Self::Dagger => vec![
                WeaponProperty::Finesse,
                WeaponProperty::Light,
                WeaponProperty::Thrown(20, 60),
            ],
            Self::Dart => vec![WeaponProperty::Finesse, WeaponProperty::Thrown(20, 60)],
            Self::Flail => vec![],
            Self::Glaive => vec![
                WeaponProperty::Heavy,
                WeaponProperty::Reach,
                WeaponProperty::TwoHanded,
            ],
            Self::Greataxe => vec![WeaponProperty::Heavy, WeaponProperty::TwoHanded],
            Self::Greatsword => vec![WeaponProperty::Heavy, WeaponProperty::TwoHanded],
            Self::Greatclub => vec![WeaponProperty::TwoHanded],
            Self::Halberd => vec![
                WeaponProperty::Heavy,
                WeaponProperty::Reach,
                WeaponProperty::TwoHanded,
            ],
            Self::Handaxe => vec![WeaponProperty::Light, WeaponProperty::Thrown(20, 60)],
            Self::Javelin => vec![WeaponProperty::Thrown(30, 120)],
            Self::Lance => vec![WeaponProperty::Reach, WeaponProperty::Special],
            Self::Longbow => vec![
                WeaponProperty::Ammunition(Ammunition::Arrows, 150, 600),
                WeaponProperty::Heavy,
                WeaponProperty::TwoHanded,
            ],
            Self::Longsword => vec![WeaponProperty::Versatile(RollCmd(1, Die::D10))],
            Self::LightHammer => vec![WeaponProperty::Light, WeaponProperty::Thrown(20, 60)],
            Self::Mace => vec![],
            Self::Maul => vec![WeaponProperty::Heavy, WeaponProperty::TwoHanded],
            Self::Morningstar => vec![],
            Self::Net => vec![WeaponProperty::Special, WeaponProperty::Thrown(5, 15)],
            Self::Pike => vec![
                WeaponProperty::Heavy,
                WeaponProperty::Reach,
                WeaponProperty::TwoHanded,
            ],
            Self::Quarterstaff => vec![WeaponProperty::Versatile(RollCmd(1, Die::D8))],
            Self::Rapier => vec![WeaponProperty::Finesse],
            Self::Scimitar => vec![WeaponProperty::Finesse, WeaponProperty::Light],
            Self::Shortbow => vec![
                WeaponProperty::Ammunition(Ammunition::Arrows, 80, 320),
                WeaponProperty::TwoHanded,
            ],
            Self::Shortsword => vec![WeaponProperty::Finesse, WeaponProperty::Light],
            Self::Sickle => vec![WeaponProperty::Light],
            Self::Sling => vec![WeaponProperty::Ammunition(
                Ammunition::SlingBullets,
                30,
                120,
            )],
            Self::Spear => vec![
                WeaponProperty::Thrown(20, 60),
                WeaponProperty::Versatile(RollCmd(1, Die::D8)),
            ],
            Self::Trident => vec![
                WeaponProperty::Thrown(20, 60),
                WeaponProperty::Versatile(RollCmd(1, Die::D8)),
            ],
            Self::WarPick => vec![],
            Self::Warhammer => vec![WeaponProperty::Versatile(RollCmd(1, Die::D10))],
            Self::Whip => vec![WeaponProperty::Finesse, WeaponProperty::Reach],
        }
    }
}

/// Types of ammunition available
#[derive(Clone, Debug, Deserialize, Display, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub enum Ammunition {
    Arrows,
    #[strum(serialize = "Blowgun needles")]
    BlowgunNeedles,
    #[strum(serialize = "Crossbow bolts")]
    CrossbowBolts,
    #[strum(serialize = "Sling bullets")]
    SlingBullets,
}

impl Ammunition {
    pub fn weapons(&self) -> Vec<Weapon> {
        match self {
            Ammunition::Arrows => vec![Weapon::Shortbow, Weapon::Longbow],
            Ammunition::BlowgunNeedles => vec![Weapon::Blowgun],
            Ammunition::CrossbowBolts => vec![
                Weapon::CrossbowLight,
                Weapon::CrossbowHand,
                Weapon::CrossbowHeavy,
            ],
            Ammunition::SlingBullets => vec![Weapon::Sling],
        }
    }
}
