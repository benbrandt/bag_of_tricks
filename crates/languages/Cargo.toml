[package]
name = "languages"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
deities = { path = "../deities" }
rand = "0.8"
rand_pcg = "0.3"
serde = { version = "1", features = ["derive"] }
strum = { version = "0.22", features = ["derive"] }
