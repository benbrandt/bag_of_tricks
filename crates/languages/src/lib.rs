use deities::{Pantheon, PantheonWeight, Pantheons};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::{Display, EnumIter, IntoEnumIterator};

#[derive(Clone, Copy, Debug, Deserialize, Display, EnumIter, PartialEq, Serialize)]
pub enum LanguageType {
    Exotic,
    Standard,
}

impl LanguageType {
    fn weight(self) -> f64 {
        match self {
            Self::Exotic => 1.0,
            Self::Standard => 2.0,
        }
    }
}

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub enum Language {
    Abyssal,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gith,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub fn gen(
        rng: &mut impl Rng,
        current_languages: &[Language],
        (amount, language_type): (usize, Option<LanguageType>),
    ) -> Vec<Self> {
        language_type
            .map_or_else(
                || Self::iter().collect::<Vec<Self>>(),
                |lt| {
                    let filtered = Self::iter()
                        .filter(|l| l.language_type() == lt)
                        .collect::<Vec<Self>>();
                    if filtered.is_empty() {
                        Self::iter().collect::<Vec<Self>>()
                    } else {
                        filtered
                    }
                },
            )
            .iter()
            .filter(|l| !current_languages.contains(l))
            .copied()
            .collect::<Vec<_>>()
            .choose_multiple_weighted(rng, amount, |l| l.language_type().weight())
            .unwrap()
            .copied()
            .collect()
    }

    fn language_type(self) -> LanguageType {
        match self {
            Self::Common
            | Self::Dwarvish
            | Self::Elvish
            | Self::Giant
            | Self::Gnomish
            | Self::Goblin
            | Self::Halfling
            | Self::Orc => LanguageType::Standard,
            Self::Abyssal
            | Self::Celestial
            | Self::DeepSpeech
            | Self::Draconic
            | Self::Gith
            | Self::Infernal
            | Self::Primordial
            | Self::Sylvan
            | Self::Undercommon => LanguageType::Exotic,
        }
    }
}

impl Pantheons for Language {
    fn addl_pantheons(&self) -> Vec<(Pantheon, PantheonWeight)> {
        (match self {
            Language::Abyssal
            | Language::Celestial
            | Language::DeepSpeech
            | Language::Gith
            | Language::Infernal
            | Language::Primordial
            | Language::Sylvan => vec![],
            Language::Common => vec![Pantheon::ForgottenRealms],
            Language::Draconic => vec![Pantheon::Dragon, Pantheon::Kobold, Pantheon::Lizardfolk],
            Language::Dwarvish => vec![Pantheon::Dwarven],
            Language::Elvish => vec![Pantheon::Elven],
            Language::Giant => vec![Pantheon::Giant],
            Language::Gnomish => vec![Pantheon::Gnomish],
            Language::Goblin => vec![Pantheon::Bugbear, Pantheon::Goblin],
            Language::Halfling => vec![Pantheon::Halfling],
            Language::Orc => vec![Pantheon::Orc],
            Language::Undercommon => vec![Pantheon::Drow, Pantheon::Duergar],
        })
        .into_iter()
        .map(|p| (p, PantheonWeight::Possible))
        .collect::<Vec<_>>()
    }
}

/// Trait for describing what languages are provided by a given object.
pub trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> (usize, Option<LanguageType>) {
        (0, None)
    }
}
