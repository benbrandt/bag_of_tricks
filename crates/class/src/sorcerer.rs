use std::fmt;

use backstory::Backstory;
use citation::{Book, Citation, CitationList, Citations};
use deities::Pantheons;
use features::Features;
use gear::weapons::WeaponType;
use languages::Languages;
use rand::Rng;
use serde::{Deserialize, Serialize};
use stats::{
    ability::{AbilityScoreType, Skill},
    equipment::StartingEquipment,
    proficiencies::{Proficiencies, Proficiency, ProficiencyOption, WeaponProficiency},
};

use super::Class;

#[derive(Default, Deserialize, Serialize)]
pub struct Sorcerer;

impl Backstory for Sorcerer {}

impl Citations for Sorcerer {
    fn citations(&self) -> CitationList {
        CitationList(vec![Citation(Book::Phb, 99)])
    }
}

impl Class for Sorcerer {
    fn gen(_: &mut impl Rng) -> Self {
        Self
    }

    fn ability_rank() -> (Vec<AbilityScoreType>, Vec<AbilityScoreType>) {
        (
            vec![AbilityScoreType::Charisma],
            vec![AbilityScoreType::Constitution],
        )
    }
}

impl Features for Sorcerer {}

impl Languages for Sorcerer {}

impl Pantheons for Sorcerer {}

impl Proficiencies for Sorcerer {
    fn proficiencies(&self) -> Vec<Proficiency> {
        vec![
            Proficiency::SavingThrow(AbilityScoreType::Charisma),
            Proficiency::SavingThrow(AbilityScoreType::Constitution),
            Proficiency::Weapon(WeaponProficiency::Specific(WeaponType::CrossbowLight)),
            Proficiency::Weapon(WeaponProficiency::Specific(WeaponType::Dagger)),
            Proficiency::Weapon(WeaponProficiency::Specific(WeaponType::Dart)),
            Proficiency::Weapon(WeaponProficiency::Specific(WeaponType::Quarterstaff)),
            Proficiency::Weapon(WeaponProficiency::Specific(WeaponType::Sling)),
        ]
    }

    fn addl_proficiencies(&self) -> Vec<ProficiencyOption> {
        vec![ProficiencyOption::Skill(
            Some(vec![
                Skill::Arcana,
                Skill::Deception,
                Skill::Insight,
                Skill::Intimidation,
                Skill::Persuasion,
                Skill::Religion,
            ]),
            2,
        )]
    }
}

impl StartingEquipment for Sorcerer {}

impl fmt::Display for Sorcerer {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Sorcerer")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let class = Sorcerer::gen(&mut rng);
        insta::assert_yaml_snapshot!(class);
    }

    #[test]
    fn test_snapshot_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let class = Sorcerer::gen(&mut rng);
        insta::assert_display_snapshot!(class);
    }

    #[test]
    fn test_ability_rank() {
        insta::assert_yaml_snapshot!(Sorcerer::ability_rank());
    }

    #[test]
    fn test_snapshot_citations() {
        let class = Sorcerer;
        insta::assert_yaml_snapshot!(class.citations());
    }

    #[test]
    fn test_snapshot_proficiences() {
        let class = Sorcerer;
        insta::assert_yaml_snapshot!(class.proficiencies());
    }

    #[test]
    fn test_snapshot_addl_proficiences() {
        let class = Sorcerer;
        insta::assert_yaml_snapshot!(class.addl_proficiencies());
    }
}
