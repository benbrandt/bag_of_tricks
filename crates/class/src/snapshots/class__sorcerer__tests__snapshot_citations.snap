---
source: crates/class/src/sorcerer.rs
expression: class.citations()

---
- - Phb
  - 99

